flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -177,6 +182,14 @@ class SubtaskCheckpointCoordinatorImpl implements SubtaskCheckpointCoordinator {
         this.checkpoints = new HashMap<>();
         this.lock = new Object();
         this.asyncOperationsThreadPool = checkNotNull(asyncOperationsThreadPool);
+        this.asyncDisposeThreadPool =
+                new ThreadPoolExecutor(
+                        0,
+                        4,
+                        60L,
+                        TimeUnit.SECONDS,
+                        new LinkedBlockingQueue<>(),
+                        new ExecutorThreadFactory("AsyncDispose"));
```

Comment:
I don't think we should be creating yet another thread pool, but just re-use the existing one. For example `asyncOperationsThreadPool` seems appropriate

Reply:
In @gaoyunhaii's comment in [FLINK-30251](https://issues.apache.org/jira/browse/FLINK-30251):

> There is also one concern:
>
> 1. The current `asyncOperationsThreadPool` is a cached thread pool, which do not have an upper limit of the number of threads, and it will create a new thread whenever there is not free thread when submitting tasks. Then if we have a large number of file to close, we might end up with a lot of threads, which might further cause a large number of memory consumption (1MB for each thread RSS region).
>
> 2. Thus we might change it to a thread pool with a limited maximum number of thread and one unbounded Blocking Queue. Also since the thread in this pool might be blocked, we might need to use a separate thread pool.

The main concern is that too many threads will be generated when asynchronous cleanup may block. From your point of view, which is better to reuse `asyncOperationsThreadPool` or create a new pool?

Reply:
In that case I would just limit the size of the `asyncOperationsThreadPool` to something like `maxConcurrentCheckpoints + 1`. It would be sensible to back pressure newer checkpoints, if system is not keeping up with deleting old checkpoints. With `maxConcurrentCheckpoints + 1` we will more or less adhere to the `maxConcurrentCheckpoints` configuration, but allow for a small leeway with allowing for simultaneous N ongoing concurrent checkpoints and clean up of one aborted one.

Actually your current design creates a risk of resource leak, if new files are created faster then old one being deleted (in case of continuously failing checkpoints).

Reply:
@pnowojski Resolved according to the comments, please review it again if you have time.